CommunityToolkit.Uwp.Authentication/WindowsProvider.cs

```diff
@@ -0,0 +1,453 @@
+using Windows.Security.Authentication.Web.Core;
+using Windows.Security.Credentials;
+using Windows.Storage;
+using Windows.UI.ApplicationSettings;
+
+namespace CommunityToolkit.Uwp.Authentication
+{
+    /// <summary>
+    /// An authentication provider based on the native AccountSettingsPane in Windows.
+    /// </summary>
+    public class WindowsProvider : BaseProvider
+    {
+        /// <summary>
+        /// Gets the redirect uri value based on the current app callback uri.
+        /// </summary>
+        public static string RedirectUri => string.Format("ms-appx-web://Microsoft.AAD.BrokerPlugIn/{0}", WebAuthenticationBroker.GetCurrentApplicationCallbackUri().Host.ToUpper());
+
+        private const string AuthenticationHeaderScheme = "Bearer";
+        private const string GraphResourcePropertyKey = "resource";
+        private const string GraphResourcePropertyValue = "https://graph.microsoft.com";
+        private const string MicrosoftAccountAuthority = "consumers";
+        private const string MicrosoftProviderId = "https://login.microsoft.com";
+        private const string SettingsKeyAccountId = "WindowsProvider_AccountId";
+        private const string SettingsKeyProviderId = "WindowsProvider_ProviderId";
+
+        // Default/minimal scopes for authentication, if none are provided.
+        private static readonly string[] DefaultScopes = { "User.Read" };
+
+        /// <summary>
+        /// Gets the client id obtained from Azure registration.
+        /// </summary>
+        public string ClientId => _clientId;
+
+        /// <summary>
+        /// Gets the list of scopes to pre-authorize during authentication.
+        /// </summary>
+        public string[] Scopes => _scopes;
+
+        /// <summary>
+        /// Gets configuration values for the AccountsSettingsPane.
+        /// </summary>
+        public AccountsSettingsPaneConfig? AccountsSettingsPaneConfig => _accountsSettingsPaneConfig;
+
+        /// <summary>
+        /// Gets a cache of important values for the signed in user.
+        /// </summary>
+        protected IDictionary<string, object> Settings => ApplicationData.Current.LocalSettings.Values;
+
+        private string _clientId;
+        private string[] _scopes;
+        private WebAccount _webAccount;
+        private AccountsSettingsPaneConfig? _accountsSettingsPaneConfig;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="WindowsProvider"/> class.
+        /// </summary>
+        /// <param name="clientId">Registered ClientId. You can access an acount without one, but any Graph requests will fail.</param>
+        /// <param name="scopes">List of Scopes to initially request.</param>
+        /// <param name="accountsSettingsPaneConfig">Configuration values for the AccountsSettingsPane.</param>
+        public WindowsProvider(string clientId = "", string[] scopes = null, AccountsSettingsPaneConfig? accountsSettingsPaneConfig = null)
```

Comment:
I assume there would also need to be a way to configure the types of accounts to sign in (personal and/or work/school) and that would be a breaking change. Consider adding that config now even if only personal accounts are supported to avoid a breaking change later

Reply:
Seeing that the clientId isn't even used for MSA accounts, I'll need to change the signature. 

@nmetulev what do you think about this, I've put the provider config details in their own object. A dev would typically NOT use this config object unless they wanted to add AAD login. But it also gives us a place to put MORE config in the future, without breaking the constructor signature for WindowsProvider:

```
[Flags]
public struct WebAccountProviderType
{
    MSA,
    AAD,
}

public struct WebAccountProviderConfig
{
    public string ClientId { get; set; }
    public WebAccountProviderType WebAccountProviderType { get; set; }
}

// Define the default
public const WebAccountProviderType DefaultWebAccountProviderType = WebAccountProviderType.MSA;

// Constructor
public WindowsProvider(string[] scopes = null, AccountsSettingsPaneConfig accountsSettingsPaneConfig = null, WebAccountProviderConfig providerConfig = null) 
{
    // If providerConfig is null, we will create one with default values:
    _providerConfig = providerConfig ?? new WebAccountsProviderConfig()
    {
        WebAccountProviderType = DefaultWebAccountProviderType,
    }
}
```

Reply:
I like it. Out of curiosity, would there be a third type for supporting both AAD and MSA, or would both types be used with bitwise operator?

Reply:
I don't think we would need a third type, but there is always room for it. My intention was to use the Flags attribute and the bitwise operator to specify one or both values (MSA/AAD).

Reply:
Oh, I almost forgot, there is potential for creating custom account providers that work with WAM. If we enable that in the future, we could add an extra value for a custom provider.